### ISSUE TITLE
[BUG] ElephasTransformer:get_config has error

### ISSUE THREAD

**Issue by OscarDPan** [maxpumperla/elephas#181]
Hi @danielenricocahall I think there's a bug in this line
https://github.com/maxpumperla/elephas/blob/master/elephas/ml_model.py#L171

`getattr(self, 'weights', [])` should be most likely returning a Broadcast object so it needs `.value` to extract and actual content

Also each weight is already `np.ndarray` so I think the `.numpy()` is weird?

This is what I ended up using but not sure if it's a perfet fix:
`'weights': [weight.tolist() for weight in self.weights.value],`

A follow-up q: do we need to make weight broadcasted before initializing ElephasTransformer?

**Comment by OscarDPan** [comment]
How about this PR https://github.com/maxpumperla/elephas/pull/182 ?
I am not sure about that failed unittest though. Doesn't look like it's related to the changes I made.

How come this save() function involves in some low-level TF stuffs?

**Comment by danielenricocahall** [comment]
I will take a look now. Thank you!

**Comment by danielenricocahall** [comment]
>>> Also each weight is already np.ndarray so I think the .numpy() is weird?


It's operating on a list of `Tensor` objects - `numpy` is the method to convert them to np arrays.

**Comment by OscarDPan** [comment]
hmm weird, `[weight.tolist() for weight in self.weights.value]` gave me error because in my application self.weights.value is a list of numpy array already.

**Comment by danielenricocahall** [comment]
Do you have a snippet to reproduce it?

**Comment by OscarDPan** [comment]
so without the PR I have to modify the get_config() a bit to get the value of the Broadcast object
`'weights': [weight.numpy().tolist() for weight in getattr(self, 'weights', []).value],`

Then it gave me this
```    'weights': [weight.numpy().tolist() for weight in getattr(self, 'weights', []).value],
AttributeError: 'numpy.ndarray' object has no attribute 'numpy'
```

Please give me sometime to make a minimal example

**Comment by OscarDPan** [comment]
Here you are
```from pyspark.ml.linalg import SparseVector
from pyspark.sql import Row, SparkSession
from pyspark.sql.types import DoubleType, StructField, ArrayType, Row
from elephas.ml_model import ElephasEstimator
from tensorflow.keras import optimizers
from tensorflow import keras
from tensorflow import keras
from tensorflow.keras.layers import Input, Dense

import numpy as np
from random import random 

def generate_multi_layer_perceptron():
    inputs = Input(shape=(3,))
    x = inputs
    x = Dense(units=1)(x)
    outputs = x
    model = keras.Model(inputs=[inputs], outputs=[outputs])
    return model

spark_session = SparkSession.builder.getOrCreate()
train_df = spark_session.createDataFrame(
    data=[
        Row(sparse_features=SparseVector(3, {1: r % 3}), blah=[3.1,2.1,4.0,0.1], blah2=[0.1], v=(1.0 if (r % 3) == 0 else 0.0)) for r in range(100)
    ]
)
keras_model = generate_multi_layer_perceptron()
opt_conf = optimizers.serialize(keras.optimizers.Adam())
estimator = ElephasEstimator(
    mode='synchronous',
    featuresCol="sparse_features",
    labelCol="v",
    outputCol="prediction",
    keras_model_config=keras_model.to_yaml(),
    categorical=False,
    num_workers=1,
    epochs=10,
    batch_size=32,
    verbose=1,
    validation_split=0.1,
    optimizer_config=opt_conf,
    loss="mse",
    predict_classes=False,
)
transformer = estimator.fit(train_df)
transformer.get_config()
```

```---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-1-8e73e9424694> in <module>
     44 )
     45 transformer = estimator.fit(train_df)
---> 46 transformer.get_config()

~/venv/lib/python3.7/site-packages/elephas/ml_model.py in get_config(self)
    169                 'labelCol': self.getLabelCol(),
    170                 'outputCol': self.getOutputCol(),
--> 171                 'weights': [weight.numpy().tolist() for weight in getattr(self, 'weights', []).value],
    172                 'model_type': getattr(self, 'model_type', None)}
    173 

~/venv/lib/python3.7/site-packages/elephas/ml_model.py in <listcomp>(.0)
    169                 'labelCol': self.getLabelCol(),
    170                 'outputCol': self.getOutputCol(),
--> 171                 'weights': [weight.numpy().tolist() for weight in getattr(self, 'weights', []).value],
    172                 'model_type': getattr(self, 'model_type', None)}
    173 

AttributeError: 'numpy.ndarray' object has no attribute 'numpy'
```

elephas==1.4.2
tensorflow==2.1.1
tensorflow-estimator==2.1.0
tensorflow-probability==0.9.0

**Comment by OscarDPan** [comment]
@danielenricocahall and on top of that, I think we should also save "featuresCol"

**Comment by OscarDPan** [comment]
Hi @danielenricocahall any update on your end? Have you been able to reproduce the error with the snippet above?

**Comment by danielenricocahall** [comment]
Yes I was able to replicate - working on it but i've been busy. Will try to have fix in by end of week.

**Comment by danielenricocahall** [comment]
Okay so now that I've had a moment to look at the problem a bit more carefully, it looks like it was just something silly in my testing - I was getting the `weights` attribute rather than calling `get_weights` method. I just pushed up the test change, along with adding the `featuresCol` to the config. Take a look over it and if it looks good to you, we can merge and tag a new patch release.

**Comment by danielenricocahall** [comment]
Thank you!

**Comment by OscarDPan** [comment]
Hi @danielenricocahall Thanks for debugging!
The changes make sense to me.

def load_ml_transformer(file_name: str):
    f = h5py.File(file_name, mode='r')
    elephas_conf = json.loads(f.attrs.get('distributed_config'), object_hook=as_enum)
    config = elephas_conf.get('config')
    return ElephasTransformer(**config)
However I think we need to do one more thing:
After saving and loading back the weight, Elephas will have attributes weights being List[float] instead of np.ndarray?

Then it would probably crash if we try to save it again since List object doesn't have tolist() method.

**Comment by danielenricocahall** [comment]
Good catch - just added a conversion step in the `load_ml_transformer` function

**Comment by OscarDPan** [comment]
yup looking good! Thanks for the work.

**Comment by danielenricocahall** [comment]
Resolved with PR https://github.com/maxpumperla/elephas/pull/182, and released on 1.4.3.
